### PR TITLE
nix: add support with upcoming 2.4 release

### PR DESCRIPTION
**Pull request by blaggacao**
Co-authored-by: David Arnold <dar@xoe.solutions>
Co-authored-by: Florian Franzen <Florian.Franzen@gmail.com>

- [x]  Change is backwards compatible.
- [x]  Commit messages are formatted correctly.
- [x]  linux & darwin tests pass.

supersedes: https://github.com/nix-community/home-manager/pull/2461
see also: https://github.com/nix-community/home-manager/issues/2438

cc @moinessim @FlorianFranzen

### CODE REVIEW

**Comment by blaggacao** [comment]
Solves:
```console
error: profile '/nix/var/nix/profiles/per-user/blaggacao/profile' is incompatible with 'nix-env'; please use 'nix profile' instead

Oops, nix-env failed to install your new Home Manager profile!

Perhaps there is a conflict with a package that was installed using
'nix-env -i'? Try running

    nix-env -q

and if there is a conflicting package you can remove it with

    nix-env -e {package name}
```

**Review by moinessim — APPROVED** [review]

**Comment by bew** [comment]
Does it automatically uses `nix profile` ? or is it opt-in in any way?

**Comment by sullyj3** [comment]
@bew It looks like it detects whether you are using `nix-env` or `nix profile` by looking at the profile (since using `nix profile` changes the format), and uses the appropriate one

**Review by rycee — COMMENTED** [review]
Thanks for the contribution! It's much appreciated. I've had a quick look and added a few comments here and there.

I'm not familiar with the new profile system and I'm a bit surprised that it is necessary to change anything outside the activation block of `home-environment.nix`. Is it not possible to use `nix-env` at all once switching to `nix profile` for the user profile? I would have guessed that `nix-env` would continue to work for managing the HM profile directory and it would only be necessary to replace the `nix-env -i` and `nix-env -e` calls.

File: modules/files.nix [hunk 1]
```diff
@@ -247,7 +247,12 @@ in
 
             local newGenFiles oldGenFiles
             newGenFiles="$(readlink -e "$newGenPath/home-files")"
-            oldGenFiles="$(readlink -e "$oldGenPath/home-files")"
+            oldGenFiles="$(readlink -e "$oldGenPath/home-files" || echo -n "")"
+
+            if [[ -z "$oldGenFiles" ]] ; then
+              echo "Did not find old home-files in $oldGenPath"
+              return
+            fi
```
[the remark this answers is not in the document]
Comment (blaggacao) [reply modules/files.nix:255]: This is coming from this commit: https://github.com/nix-community/home-manager/pull/2461/commits/a9aaf38a11aa7a5323cc72e92cf49d92e1fd89be

So I'd assume that it is a problem on first install or so.

**Comment by blaggacao** [comment]
> Is it not possible to use nix-env at all once switching to nix profile for the user profile?

Indeed it is _not_ possible at all to use `nix-env` at all in that scenario.

I have adopted your changes and squashed them.

**Comment by rycee** [comment]
@blaggacao Thanks for the clarification. Then I guess we really have to do a wholesale move-over.

I'm doing some additional cleanups in PR #2840 and will merge from there so will close this PR.

**Pull request closed without merging**